File: semester_3/operating_system/src/rust/pthread/src/bin/pthread-cond.rs
/*
 * pthread/src/bin/pthread-cond.rs --
 *
 *      A simple demonstration of condition variables in Rust. This
 *      example lacks error handling code to keep it short.
 */

use std::env;
use std::sync::{Arc, Condvar, Mutex};
use std::thread;

struct Counter {
    mutex: Mutex<u64>,
    cvar: Condvar,
}

fn even(arc: Arc<Counter>) {
    let mut ctr = arc
        .cvar
        .wait_while(arc.mutex.lock().unwrap(), |c| *c % 2 != 0)
        .unwrap();
    eprintln!("even: {}", *ctr);
    *ctr += 1;
    arc.cvar.notify_all();
}

fn odd(arc: Arc<Counter>) {
    let mut ctr = arc
        .cvar
        .wait_while(arc.mutex.lock().unwrap(), |c| *c % 2 == 0)
        .unwrap();
    eprintln!("odd:  {}", *ctr);
    *ctr += 1;
    arc.cvar.notify_all();
}

fn main() {
    let ctr = Counter {
        mutex: Mutex::new(0),
        cvar: Condvar::new(),
    };
    let data = Arc::new(ctr);

    let mut vec = Vec::new();
    for _ in env::args().skip(1) {
        let c = data.clone();
        vec.push(thread::spawn(move || even(c)));
        let c = data.clone();
        vec.push(thread::spawn(move || odd(c)));
    }

    for handle in vec {
        handle.join().unwrap();
    }
}

File: semester_3/operating_system/src/c/pthread/pthread-cond.c
/*
 * pthread/pthread-cond.c --
 *
 *      A simple demonstration of pthread condition variables. This
 *      example lacks error handling code to keep it short.
 */

#define _POSIX_C_SOURCE 200809L

#include <pthread.h>

typedef struct {
    unsigned int counter;       /* shared counter */
    pthread_mutex_t mutex;      /* mutex protecting the counter */
    pthread_cond_t cond;        /* condition variable */
} counter_t;

static void* even(void *arg)
{
    counter_t *c = (counter_t *) arg;
    (void) pthread_mutex_lock(&c->mutex);
    while (c->counter % 2 != 0) {
        (void) pthread_cond_wait(&c->cond, &c->mutex);
    }
    c->counter++;
    (void) pthread_mutex_unlock(&c->mutex);
    (void) pthread_cond_signal(&c->cond);
    return NULL;
}

static void* odd(void *arg)
{
    counter_t *c = (counter_t *) arg;
    (void) pthread_mutex_lock(&c->mutex);
    while (c->counter % 2 == 0) {
        (void) pthread_cond_wait(&c->cond, &c->mutex);
    }
    c->counter++;
    (void) pthread_mutex_unlock(&c->mutex);
    (void) pthread_cond_signal(&c->cond);
    return NULL;
}

int main(int argc, char *argv[])
{
    pthread_t tids[2*argc];
    counter_t cnter = { .counter = 0, .mutex = PTHREAD_MUTEX_INITIALIZER,
                        .cond = PTHREAD_COND_INITIALIZER };
    (void) argv;
    
    for (int i = 1; i < argc; i++) {
        (void) pthread_create(&tids[2*i], NULL, even, &cnter);
        (void) pthread_create(&tids[2*i+1], NULL, odd, &cnter);
    }
    for (int i = 1; i < argc; i++) {
        (void) pthread_join(tids[2*i], NULL);
        (void) pthread_join(tids[2*i+1], NULL);
    }
    return 0;
}
